### ISSUE TITLE
about circular dependency

### ISSUE THREAD

**Issue by Wmerake** [cbeust/testng#1874]
### TestNG Version
We tried in latest Version TestNG -6.14.0 . Expected behaviour is not supported yet.
Please reconfirm

### Expected behavior
run with the ordered test case below
![image](https://user-images.githubusercontent.com/13835804/43087697-099489be-8ed3-11e8-83d2-17af948bbbe1.png)

### Actual behavior
but testng gives me the exception about
```
java.lang.IllegalStateException: Circular dependency: AcctMqReceiverTest.receive(java.lang.String)[pri:0, instance:com.qianmi.d2c.test.custom.api.mq.receiver.AcctMqReceiverTest@3059cbc] <-> D2CTradeSellerProviderTest.create(java.lang.String, java.lang.String)[pri:0, instance:com.qianmi.d2c.test.generated.parameterized.api.trade.seller.D2CTradeSellerProviderTest@6babf3bf]
	at org.testng.internal.DynamicGraph.addEdges(DynamicGraph.java:83)
	at org.testng.internal.DynamicGraph.addEdge(DynamicGraph.java:58)
	at org.testng.TestRunner.createDynamicGraph(TestRunner.java:1103)
	at org.testng.TestRunner.privateRun(TestRunner.java:747)
	at org.testng.TestRunner.run(TestRunner.java:634)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:425)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:420)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:385)
	at org.testng.SuiteRunner.run(SuiteRunner.java:334)
```

so I change the sortMethod method in MethodHelper and sortMethodsByInstance method in class MethodHelper to make the test case run by the order in test case. and then it's running OK
![image](https://user-images.githubusercontent.com/13835804/43114708-dfe75c94-8f32-11e8-9c1a-48b93c034baa.png)

and I think if the  circular dependency is necessary to our test case ? in my project there's about 10000
test case running ,to avoid the circular dependency in test case .We need to make another test class though they are one.
like this  
![image](https://user-images.githubusercontent.com/13835804/43114573-52b5b71c-8f32-11e8-9b90-d2bcadfa7a4d.png)

So I think there's no necessary to check the  circular dependency.
I've read almost the whole code in your program, as my pleasure.

### Is the issue reproductible on runner?

- [ ] Shell
- [ ] Maven
- [ ] Gradle
- [ ] Ant
- [ ] Eclipse
- [x] IntelliJ
- [ ] NetBeans

### Test case sample
> Please, share the test case (as small as possible) which shows the issue

**Comment by krmahadevan** [comment]
@Wmerake - I am not quite sure I understand the issue here. Would it be possible for you to please help share a sample that can be used to reproduce the problem ? The screenshots aren't helping a lot in understanding what your tests look like and what is perhaps causing the exception. 

It looks like you have a dependency defined in your methods and you perhaps may also have some priority defined ?

**Comment by Wmerake** [comment]
yeap,my test case need to run with the sort in my test case.
[sample.txt](https://github.com/cbeust/testng/files/2222160/sample.txt)

**Comment by krmahadevan** [comment]
Not sure if I was clear. Let me rephrase.

Please show us what the source code for the following look like :

1.
com.qianmi.d2c.test.generated.parameterized.api.trade.seller.D2CTradeSellerProviderTest.create()
2. com.qianmi.d2c.test.custom.api.mq.receiver.AcctMqReceiverTest.receive()
3. com.qianmi.d2c.test.generated.parameterized.api.trade.seller.D2CTradeSellerProviderTest.dispatch()

If it has a lot of external dependencies, please help trim down these
methods such that it has only sysouts so that we can run it easily and
reproduce the problem.

**Comment by Wmerake** [comment]
here's the test case demo about circular dependency.
[testcase.zip](https://github.com/cbeust/testng/files/2227705/testcase.zip)
